src/main/scala/dxWDL/compiler/Native.scala

```diff
@@ -106,40 +106,85 @@ case class Native(dxWDLrtId: Option[String],
+          case IR.IOAttrPatterns(patternRepr) =>
+            patternRepr match {
+              case IR.PatternsReprArray(patterns) =>
+                Some(IR.PARAM_META_PATTERNS -> JsArray(patterns.map(JsString(_))))
+              // If we have the alternative patterns object, extrac the values, if any at all
+              case IR.PatternsReprObj(name, klass, tags) =>
+                val attrs: Map[String, JsValue] = List(
+                    if (name.isDefined) Some("name" -> JsArray(name.get.map(JsString(_))))
+                    else None,
+                    if (tags.isDefined) Some("tag" -> JsArray(tags.get.map(JsString(_))))
+                    else None,
+                    if (klass.isDefined) Some("class" -> JsString(klass.get)) else None
+                ).flatten.toMap
+                // If all three keys for the object version of patterns are None, return None
+                if (attrs.isEmpty) None else Some(IR.PARAM_META_PATTERNS -> JsObject(attrs))
+            }
+          case IR.IOAttrChoices(choices) =>
+            Some(IR.PARAM_META_CHOICES -> JsArray(choices.map(choice => {
+              choice match {
+                case IR.ChoiceReprString(value) => JsString(value)
+                case IR.ChoiceReprInteger(value) => JsNumber(value)
+                case IR.ChoiceReprFloat(value) => JsNumber(value)
+                case IR.ChoiceReprBoolean(value) => JsBoolean(value)
+                case IR.ChoiceReprFile(value, name) => {
+                  // TODO: support project and record choices
+                  val dxLink = DxPath.resolveDxURLFile(value).getLinkAsJson
+                  if (name.isDefined) {
+                    JsObject(Map("name" -> JsString(name.get), "value" -> dxLink))
+                  } else {
+                    dxLink
+                  }
+                }
               }
-            case _ => None
-          }.toMap
-        }
+            })))
+          case IR.IOAttrSuggestions(suggestions) =>
+            Some(IR.PARAM_META_SUGGESTIONS -> JsArray(suggestions.map(suggestion => {
+              suggestion match {
+                case IR.SuggestionReprString(value) => JsString(value)
+                case IR.SuggestionReprInteger(value) => JsNumber(value)
+                case IR.SuggestionReprFloat(value) => JsNumber(value)
+                case IR.SuggestionReprBoolean(value) => JsBoolean(value)
+                case IR.SuggestionReprFile(value, name, project, path) => {
+                  // TODO: support project and record suggestions
+                  val dxLink: Option[JsValue] = value match {
+                    case Some(str) => Some(DxPath.resolveDxURLFile(str).getLinkAsJson)
+                    case None => None
+                  }
+                  if (name.isDefined || project.isDefined || path.isDefined) {
+                    val attrs: Map[String, JsValue] = List(
+                      if (dxLink.isDefined) Some("value" -> dxLink.get) else None,
```

Comment:
This a nice pattern for dealing with multiple elements that are optional.

Reply:
I learned it from @sstadick  :)